psrecord/main.py

```diff
@@ -45,19 +47,20 @@ def get_memory(process):
 
 
 def all_children(pr):
-    processes = []
-    children = []
+    global children
+
     try:
-        children = pr.children()
+        children_of_pr = pr.children()
     except AttributeError:
-        children = pr.get_children()
+        children_of_pr = pr.get_children()
     except Exception:  # pragma: no cover
         pass
 
-    for child in children:
-        processes.append(child)
-        processes += all_children(child)
-    return processes
+    for child in children_of_pr:
+        if child not in children:
+            children.append(child)
+            all_children(child)
```

Comment:
I think the recursive call `all_children(child)` should be outside of the conditional: with the current code, if child a process launches new children after it was first encountered, then those new children will not be included.

Reply:
No longer relevant since I've added a commit to delegate the recursive calling to psutil